### DISCUSSION TITLE
@- org standard for anti-hijacking/substitution

### DISCUSSION THREAD

**Discussion by Tymek** [npm/feedback#366] [Registry] [answered]
I would like to suggest converting [`@-`](https://www.npmjs.com/org/-) org for use in local packages (monorepositories), safe from hijacking. I have `@-local` for this.

- It's reserved, but not in use.
- Allowed by package name convention `^(?:@[a-z0-9-*~][a-z0-9-*._~]*/)?[a-z0-9-~][a-z0-9-._~]*$`
- Short, simple and clean. Easier than for example `@localrepo`
- Is on top of A-Z sorted `dependencies`
- Just looks right. For example: `@-/package`, `@-/client-app`, `@-/monorepository`

If this proposal is not actionable, maybe adding `@/package` is?

**Comment by ljharb** [comment]
`@` can't be a package name, so you can safely use `@/foo` for that. You can also use other characters that can't be in a package name - `~`, `^`, etc.

**Reply by Tymek** [reply]
Sure. I have (and use) `@-local`. I'd like to use allowed package name. Doing something different will become problematic very quickly (CI/CD, scripts, bots etc).  Issue here is a lack of standard, and `@-` could be great IMHO 🙂 

![Screenshot from 2021-05-03 19-57-27](https://user-images.githubusercontent.com/2625371/116913482-d3be8280-ac49-11eb-9dde-1e0450554579.png)

**Comment by MylesBorins** [comment, marked as answer]
Hey @tymek thanks for the suggestion. I guess there are two different was to break apart this suggestion.

1. What type of CLI changes would need to be made to support this. The answer to this might potentially be node
2. What type of changes we might need to make to the registry. Seems like we might need to make a hold on this scope... this would guarantee that there are no public packages that could utilize this scope to publish, but it would not guarantee that other registries would not allow this scope to be published to (which in some cases is to be expected).

**Reply by Tymek** [reply]
To be honest, I didn't ponder about it much. Substitution or lack of standard for local-only scope can be a problem with official public registry, I don't understand vectors or problems with private ones. I'm not sure what type of change would be the best. My hope is to have _recommended_ way of doing local-only monorepo, not necessarily new _standard_ or breaking change. 

Inspiration for this where _pseudo-top-level domains_ on the web: `.local` and `.localhost`

**Reply by Tymek** [reply]
:thinking:

`@-` is not in use. I would like it to be used like [@localrepo/pledge](https://www.npmjs.com/package/@localrepo/pledge). On `@-/-`. According to [docs.npmjs.com/cli/v6/using-npm/disputes](https://docs.npmjs.com/cli/v6/using-npm/disputes), can I file a dispute with a claim of squatting :sweat_smile:?

**Reply by ljharb** [reply]
That seems like a lot of work rather than just using `@/` or another unusable scope name.

**Reply by evisong** [reply]
It's 2022, and more hijacking happended in past months. I think the standards referred by @SchnWalter in https://github.com/npm/npm/issues/20680 can be good candidates.

> * RFC2606 - [3. Reserved Example Second Level Domain Names](https://tools.ietf.org/html/rfc2606#section-3):
>    * example.com
>    * example.org
> * RFC6761 - [6.3. Domain Name Reservation Considerations for "localhost."](https://tools.ietf.org/html/rfc6761#section-6.3):
>    * localhost.site
>    * example.localhost.site

> - `@app`
> - `@example`
> - `@local`
> - `@private`
> - `@shared`
> - `@src`

npmjs.org is the most commonly used public upstream of all private repos, it will be great if npm could drive this.

**Reply by ljharb** [reply]
What “hijacking” are you referring to? It’s always been a best practice to install from a private registry, and to use a scope for internal packages that you own publicly. This proposal would obviate the need for the latter, but as i said, that’s a lot of complexity to build just to avoid “picking and registering a public scope” to reserve it for your internal use.

**Reply by SchnWalter** [reply]
I've been tagged, so I'll chip in.

--- 

> @Tymek: […] safe from hijacking

Can you please explain what's your threat model here? My guess is that you're talking about the threat presented in the following CWE:

> [CWE-427: Uncontrolled Search Path Element](https://cwe.mitre.org/data/definitions/427.html)
> 
> The product uses a fixed or controlled search path to find resources, but one or more locations in that path can be under the control of unintended actors.  
> 
> In software package management frameworks (e.g., npm, RubyGems, or PyPi), the framework may identify dependencies on third-party libraries or other packages, then consult a repository that contains the desired package. The framework may search a public repository before a private repository. This could be exploited by attackers by placing a malicious package in the public repository that has the same name as a package from the private repository. The search path might not be directly under control of the developer relying on the framework, but this search order effectively contains an untrusted element.   

If that's the case, I'm afraid that introducing such a change is not worth it, because the new naming pattern will break a lot of code out there in the wild. You've mentioned that this is a "local-only" change, it's not.

Having reserved namespaces can be useful, but not for the reasons mentioned in this issue. Also, there's no way to stop developers from shooting themselves in the foot. They are on their own if they make spelling errors or start using unreserved names, check the "Potential Mitigations" section in the linked CWE.

--- 

> @Tymek: […] for use in local packages

Can you please explain how you define those "local packages"? 

1) If you are talking about packages contain a [local-path](https://docs.npmjs.com/cli/v8/configuring-npm/package-json/#local-paths), then there is no threat model here, since you will always use the local package, regardless if it exists in a remote registry or not.

2) If you are talking about packages that exist in your private registry, then it makes no to sens to define those as "local packages", it will only cause confusion.

--- 

> @evisong: It's 2022, and more hijacking happended in past months 

I'm guessing that you talk about those wrongly titled "NPM library hijacked" posts.  Those are unrelated, and we've got some improvements after that mess: [package maintainers now require 2FA](https://github.blog/2021-12-07-enrolling-npm-publishers-enhanced-login-verification-two-factor-authentication-enforcement/). In that case, you've nothing to worry about.

--- 

These being said, the reasoning behind https://github.com/npm/npm/issues/20680 is still valid today. New users will always copy/paste from documentation. Having some reserved namespaces for documentation purposes, especially those specified from the linked RFCs, will offer some minimal protection against bad actors that register the example packages. I'll submit a new issue to [npm/rfcs](https://github.com/npm/rfcs) repo.

**Reply by evisong** [reply]
> It’s always been a best practice to install from a private registry, and to use a scope for internal packages that you own publicly.

@ljharb Thanks for the quick response. 

I totally agree with you. It's important for an organization to take control of its own artifacts.

In my company, we do share a central private npm registry on top of JFrog Artifactory. And for historical reasons, we have some internal npm packages published in this private registry, whose package names were not carefully picked, some are even not scoped, e.g. `@department-a/lib-1`, `department-b-lib-2`.

When setting up every new dev environment or CI/CD environment, the first thing to do after installing Node and npm, is to create a `~/.npmrc` and configure this private registry as default instead of the public `npmjs.com`; otherwise, running `npm install` or `npx` will lead to unexpected results, if the package happens to exist in the public registry but without our awareness.

To mitigate such risk, there're 2 options:

1. Collect all existing internal scope names and unscoped package names, register them in `npmjs.com`;
2. Register a company-owned scope in `npmjs.com`, rename all internal package names with this scope and republish.

Actually I don't think it's feasible to adjust npm to handle these historical complexities, instead, it would be simple if `npmjs.com` defines some reserved scopes, e.g. `@example`, `@demo`, etc. for every new npm user or environment, installing/publishing packages by using reserved scopes would **fail fast**, and the user will be prompted: 

```sh
$ npm install @example/my-company-lib-1
Error: The `@example` scope is only valid in a private registry, please config one in `npm config` or `~/.npmrc` first.
```

By introducing the mechanism above, by default, an npm user could safely publish internal packages without the needs for registering new "placeholder" scopes in `npmjs.com`.

**Reply by ljharb** [reply]
I agree that those indeed are your two options (i'd suggest the second, but either works).

Regarding your other suggestion, npm isn't the only public registry, so I don't think such a reservation makes sense.

**Reply by travi** [reply]
> When setting up every new dev environment or CI/CD environment, the first thing to do after installing Node and npm, is to create a `~/.npmrc` and configure this private registry as default instead of the public `npmjs.com`

including a `.npmrc` in the project prevents the need for this to be done manually by each developer

**Reply by evisong** [reply]
> the reasoning behind [npm/npm#20680](https://github.com/npm/npm/issues/20680) is still valid today.

@SchnWalter I just realize I'm mixing several topics together, sorry for the mess.  But I do appreciate the ideas you posted in [npm/npm#20680](https://github.com/npm/npm/issues/20680), they're inspiring me.
